File: backend/app/api/v1/promotions.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.core.security import get_current_user, require_roles
from app.models.user import User
from app.schemas.promotion import PromotionOut, PromotionSave
from app.services import promotion_service

router = APIRouter(prefix="/promotions", dependencies=[Depends(get_current_user)])


@router.get("/active", response_model=list[PromotionOut])
def list_active_promotions(db: Session = Depends(get_db)):
    """Promotions running today — shown to branches as labels by the product name."""
    return promotion_service.list_active(db)


@router.get("", response_model=list[PromotionOut])
def list_promotions(
    _admin: User = Depends(require_roles("ADMIN")),
    db: Session = Depends(get_db),
):
    """Every saved promotion, including upcoming and ended ones."""
    return promotion_service.list_all(db)


@router.put("", response_model=list[PromotionOut])
def save_promotions(
    payload: PromotionSave,
    admin: User = Depends(require_roles("ADMIN")),
    db: Session = Depends(get_db),
):
    """Replaces the whole list; an empty list clears every promotion."""
    return promotion_service.save_all(db, admin, payload)

File: frontend/src/api/promotions.ts
import { apiFetch } from "./client";

export type PromotionType = "FRESH_CHOICE" | "SPECIAL_WEEKEND" | "SPECIAL";

export type Promotion = {
  product_id: number;
  promotion_type: PromotionType;
  start_date: string; // YYYY-MM-DD, inclusive
  end_date: string; // YYYY-MM-DD, inclusive
};

// Promotions running today — what branches see as labels.
export function fetchActivePromotions(): Promise<Promotion[]> {
  return apiFetch("/promotions/active");
}

// Admin only: every saved promotion, including upcoming and ended ones.
export function fetchAllPromotions(): Promise<Promotion[]> {
  return apiFetch("/promotions");
}

// Admin only: replaces the whole list — an empty list clears everything.
export function savePromotions(lines: Promotion[]): Promise<Promotion[]> {
  return apiFetch("/promotions", {
    method: "PUT",
    body: JSON.stringify({ lines }),
  });
}
